EFI & Secure Boot

stejskalleos opened the pull request.

ekohl reviewed (changes requested):
Please don't change formatting in the same commit. Also, what's different than https://github.com/fog/fog-libvirt/pull/134? Can't that be used/merged?

stejskalleos commented:
Yeah that's my editor settings, need to disable it.

> Also, what's different than https://github.com/fog/fog-libvirt/pull/134? Can't that be used/merged?

I can't edit current PRs/branches of others, so I thought I would create a new one since we agreed that I would continue with the work.

stejskalleos commented:
Tests are failing on master as well, don't know the details yet.

ekohl commented:
> I can't edit current PRs/branches of others, so I thought I would create a new one since we agreed that I would continue with the work.

I said that I thought the fog-libvirt patch was OK, but should only be merged once we have the Foreman code in a mergeable state so we know it's a good API. Taking a commit from someone else without attributing the original author is a poor practice that leans towards plagiarism. Even if it's open source, crediting the original author is important. And if there is prior work, explaining why your version is different.

> Tests are failing on master as well, don't know the details yet.

#140

stejskalleos commented:
> crediting the original author is important. 

Yeah I didn't want to steal your work, I can credit you for sure, just wanted to make it fast as possible.

stejskalleos commented:
@ekohl I assigned you in the https://github.com/theforeman/foreman/pull/10209/files wym to take care of the review ?

stejskalleos commented:
Rebased, added @ekohl as author & tested with Fedora 39 UEFI + SecureBoot. Ready for review
Foreman PR: https://github.com/theforeman/foreman/pull/10209

jloeser commented:
We should implement it according to https://libvirt.org/kbase/secureboot.html:

Enable SB:
```<os firmware='efi'>
  <firmware>
    <feature enabled='yes' name='secure-boot'/>
    <feature enabled='yes' name='enrolled-keys'/>
  </firmware>
</os>

```
Disable SB:
```
<os firmware='efi'>
  <firmware>
    <feature enabled='no' name='secure-boot'/>
  </firmware>
</os>
```

Providing `<loader secure="yes|no"/>` is not sufficient (see https://github.com/theforeman/foreman/pull/10209#discussion_r1648583006).

davispuh commented:
Looking forward to this getting merged.

ekohl commented on tests/libvirt/models/compute/server_tests.rb:
```diff
@@ -79,5 +82,45 @@
+      xml = server.to_xml
+
+      os_firmware = xml.include?('<os firmware="efi">')
+      secure_boot = !xml.include?('<feature name="secure-boot" enabled="no" />')
+      enrolled_keys = !xml.include?('<feature name="enrolled-keys" enabled="no" />')
```
I think this isn't a good test because there's a very high chance the feature is just slightly different. In particular, when I run this test manually I see it includes:
```xml
    <firmware>
      <feature name="secure-boot" enabled="no"/>
      <feature name="enrolled-keys" enabled="no"/>
    </firmware>
```

Note the space before `/>` is missing.

Did you intend:
```suggestion
      secure_boot = xml.include?('<feature name="secure-boot" enabled="no"/>')
      enrolled_keys = xml.include?('<feature name="enrolled-keys" enabled="no"/>')
```

ekohl commented on lib/fog/libvirt/models/compute/server.rb:
```diff
@@ -281,14 +283,28 @@ def to_xml
                   xml.boot(:dev => dev)
                 end
+
+                if os_firmware == "efi"
```
Should this be:
```suggestion
                if os_firmware == "efi" && os_firmware_features.any?
```

ekohl replied:
Also, please make sure we deal with `nil`
```suggestion
                if os_firmware == "efi" && os_firmware_features&.any?
```

nofaralfasi commented:
I don't have permission to push changes to this PR, so I created a new one: https://github.com/fog/fog-libvirt/pull/155.
